### PR TITLE
beacon/engine, eth/catalyst: EIP-4844 updates for the engine API

### PR DESCRIPTION
**Pull request by fjl**
This is a spin-out from the EIp-4844 development branch, containing just the Engine API modifications
and nothing else. The newPayloadV3 endpoint won't really work in this version, but we need the data structures
for testing so I'd like to get this in early.

### CODE REVIEW

File: eth/catalyst/api.go [hunk 1]
```diff
@@ -405,23 +407,13 @@ func (api *ConsensusAPI) GetPayloadV2(payloadID engine.PayloadID) (*engine.Execu
...
 }
 
-func (api *ConsensusAPI) getPayload(payloadID engine.PayloadID) (*engine.ExecutionPayloadEnvelope, error) {
-	log.Trace("Engine API request received", "method", "GetPayload", "id", payloadID)
-	data := api.localBlocks.get(payloadID, false)
-	if data == nil {
-		return nil, engine.UnknownPayload
-	}
-	return data, nil
+// GetPayloadV2 returns a cached payload by id.
```
Comment (MariusVanDerWijden) [inline eth/catalyst/api.go:410]:
```suggestion
// GetPayloadV3 returns a cached payload by id.
```

File: eth/catalyst/api.go [hunk 2]
```diff
@@ -405,23 +407,13 @@ func (api *ConsensusAPI) GetPayloadV2(payloadID engine.PayloadID) (*engine.Execu
...
 
-// getFullPayload returns a cached payload by it. The difference is that this
-// function always expects a non-empty payload, but can also return empty one
-// if no transaction is executable.
-//
-// Note, this function is not a part of standard engine API, meant to be used
-// by consensus client mock in dev mode.
-func (api *ConsensusAPI) getFullPayload(payloadID engine.PayloadID) (*engine.ExecutionPayloadEnvelope, error) {
-	log.Trace("Engine API request received", "method", "GetFullPayload", "id", payloadID)
+func (api *ConsensusAPI) getPayload(payloadID engine.PayloadID) (*engine.ExecutionPayloadEnvelope, error) {
```
Comment (MariusVanDerWijden) [inline eth/catalyst/api.go:415]: This removes some code that is used by the new devmode afaik

File: eth/catalyst/api.go [hunk 3]
```diff
@@ -730,8 +741,8 @@ func (api *ConsensusAPI) ExchangeCapabilities([]string) []string {
 	return caps
 }
 
-// GetPayloadBodiesByHashV1 implements engine_getPayloadBodiesByHashV1 which
-// allows for retrieval of a list of block bodies by the engine api.
+// GetPayloadBodiesV1 implements engine_getPayloadBodiesByHashV1 which allows for retrieval of a list
```
Comment (MariusVanDerWijden) [inline eth/catalyst/api.go:744]:
```suggestion
// GetPayloadBodiesByHashV1 implements engine_getPayloadBodiesByHashV1 which allows for retrieval of a list
```

**Comment by fjl** [comment]
@MariusVanDerWijden PTAL

**Comment by fjl** [comment]
Rebased after removal of les/catalyst.

**Pull request merged**